src/controller/src/clusters.rs

```diff
@@ -190,46 +190,92 @@ where
 
     /// Creates a replica of the specified cluster with the specified identifier
     /// and configuration.
-    pub async fn create_replica(
+    ///
+    /// This method is NOT idempotent; It can fail between processing of different
+    /// replicas and leave the controller in an inconsistent state. It is almost
+    /// always wrong to do anything but abort the process on `Err`.
+    pub async fn create_replicas(
         &mut self,
-        cluster_id: ClusterId,
-        replica_id: ReplicaId,
-        config: ReplicaConfig,
+        replicas: Vec<(ClusterId, ReplicaId, ReplicaConfig)>,
     ) -> Result<(), anyhow::Error> {
-        let (storage_addr, compute_location) = match config.location {
-            ReplicaLocation::Unmanaged(u) => {
-                let storage_addr = u.storagectl_addr;
-                let compute_location = ComputeReplicaLocation {
-                    computectl_addrs: u.computectl_addrs,
-                    compute_addrs: u.compute_addrs,
-                    workers: u.workers,
-                };
-                (storage_addr, compute_location)
-            }
-            ReplicaLocation::Managed(m) => {
-                let workers = m.allocation.workers;
-                let service = self.provision_replica(cluster_id, replica_id, m).await?;
-                let storage_addr = service
-                    .addresses("storagectl")
-                    .into_iter()
-                    .next()
-                    .expect("should be at least one process");
-                let compute_location = ComputeReplicaLocation {
-                    computectl_addrs: service.addresses("computectl"),
-                    compute_addrs: service.addresses("compute"),
-                    workers,
-                };
-                (storage_addr, compute_location)
-            }
-        };
+        // Reborrow the `&mut self` as immutable, as all the concurrent work to be processed in
+        // this stream cannot all have exclusive access.
+        let this = &*self;
+        let replicas: Vec<_> = futures::stream::iter(replicas)
```

Comment:
I'm curious about why you choose using a stream instead of spawning individual tasks and `join`ing them all (which would have been my naive approach).

Reply:
@teskje because `buffer_unordered` is _much_ a much easier and less manual way to do that, while also limiting the max amount of concurrency!